linear_algebra/src/conjugate_gradient.py

```diff
@@ -0,0 +1,176 @@
+import numpy as np
+
+# https://en.wikipedia.org/wiki/Conjugate_gradient_method
+# https://en.wikipedia.org/wiki/Definite_symmetric_matrix
+
+
+def _is_matrix_spd(matrix: np.array) -> bool:
+
+    """
+    Returns True if input matrix is symmetric positive definite.
+    Returns False otherwise.
+
+    For a matrix to be SPD, all eigenvalues must be positive.
+
+    >>> import numpy as np
+    >>> matrix = np.array([
+    ... [4.12401784, -5.01453636, -0.63865857],
+    ... [-5.01453636, 12.33347422, -3.40493586],
+    ... [-0.63865857, -3.40493586,  5.78591885]])
+    >>> _is_matrix_spd(matrix)
+    True
+    >>> matrix = np.array([
+    ... [0.34634879,  1.96165514,  2.18277744],
+    ... [0.74074469, -1.19648894, -1.34223498],
+    ... [-0.7687067 ,  0.06018373, -1.16315631]])
+    >>> _is_matrix_spd(matrix)
+    False
+    """
+    # Ensure matrix is square.
+    assert np.shape(matrix)[0] == np.shape(matrix)[1]
+
+    # If matrix not symmetric, exit right away.
+    if np.allclose(matrix, matrix.T) is False:
+        return False
+
+    # Get eigenvalues and eignevectors for a symmetric matrix.
+    eigen_values, _ = np.linalg.eigh(matrix)
+
+    # Check sign of all eigenvalues.
+    return np.all(eigen_values > 0)
+
+
+def _create_spd_matrix(N: np.int64) -> np.array:
+    """
+    Returns a symmetric positive definite matrix given a dimension.
+
+    Input:
+    N gives the square matrix dimension.
+
+    Output:
+    spd_matrix is an NxN symmetric positive definite (SPD) matrix.
+
+    >>> import numpy as np
+    >>> N = 3
+    >>> spd_matrix = _create_spd_matrix(N)
+    >>> _is_matrix_spd(spd_matrix)
+    True
+    """
+
+    random_matrix = np.random.randn(N, N)
+    spd_matrix = np.dot(random_matrix, random_matrix.T)
+
+    assert _is_matrix_spd(spd_matrix)
+
+    return spd_matrix
+
+
+def conjugate_gradient(
+    spd_matrix: np.array, b: np.array, max_iterations=1000, tol=1e-8
+) -> np.array:
+    """
+    Returns solution to the linear system np.dot(spd_matrix, x) = b.
+
+    Input:
+    spd_matrix is an NxN Symmetric Positive Definite (SPD) matrix.
+    b is an Nx1 vector that is the load vector.
+
+    Output:
+    x is an Nx1 vector that is the solution vector.
+
+    >>> import numpy as np
+    >>> spd_matrix = np.array([
+    ... [8.73256573, -5.02034289, -2.68709226],
+    ... [-5.02034289,  3.78188322,  0.91980451],
+    ... [-2.68709226,  0.91980451,  1.94746467]])
+    >>> b = np.array([
+    ... [-5.80872761],
+    ... [ 3.23807431],
+    ... [ 1.95381422]])
+    >>> conjugate_gradient(spd_matrix,b)
+    array([[-0.63114139],
+           [-0.01561498],
+           [ 0.13979294]])
+    """
+    # Ensure proper dimensionality.
+    assert np.shape(spd_matrix)[0] == np.shape(spd_matrix)[1]
+    assert np.shape(b)[0] == np.shape(spd_matrix)[0]
+    assert _is_matrix_spd(spd_matrix)
+
+    N = np.shape(b)[0]
+
+    # Initialize solution guess, residual, search direction.
+    x0 = np.zeros((N, 1))
```

Comment:
When you create a variable and then use that variable only once on the very next line then try to remove that variable unless 1. The second line is _very_ long or 2. the variable name documents something that is unclear.

`N` in this context does neither.

```suggestion
    # Initialize solution guess, residual, search direction.
    x0 = np.zeros((np.shape(b)[0], 1))
```